File: services/common/companions.py
"""Stable, deliberately light companion definitions.

Companion styles are product delivery settings, never a source of an account
owner's Persona or Digital Self material.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from typing import Final, Literal

DEFAULT_COMPANION_ID: Final = "starlight"
DESIGNED_VOICE_MODEL: Final = "seed-tts-2.0"

COMPANION_STYLE_VERSION: Final = "companion-v1"


@dataclass(frozen=True, slots=True)
class CompanionDefinition:
    """The complete permitted companion surface from ADR-0016."""

    companion_id: str
    display_name: str
    style_description: str
    warmth: str
    directness: str
    response_length: str
    question_frequency: str
    interview_depth: str
    designed_voice_profile: str
    welcome_text: str
    conversation_instruction: str
    voice_instruction: str
    default_voice_emotion: Literal["neutral", "happy"]
    default_voice_rate: float


COMPANIONS: Final[dict[str, CompanionDefinition]] = {
    "starlight": CompanionDefinition(
        "starlight",
        "星澜",
        "温暖回应，偶尔陪用户把想法理清一层",
        "warm",
        "gentle",
        "balanced",
        "occasional",
        "light",
        "warm_companion",
        "嗨，我是星澜。今天想聊点什么，我陪你慢慢说。",
        "保持温暖、从容和有分寸的好奇；先承接用户最在意的点，再偶尔帮他把想法理清一层。",
        "声音温暖明亮，语速略慢且从容，停顿自然，不要甜腻或客服腔。",
        "neutral",
        0.99,
    ),
    "taoxi": CompanionDefinition(
        "taoxi",
        "桃喜",
        "轻快回应，回复偏短，只做少量追问",
        "bright",
        "direct",
        "brief",
        "occasional",
        "light",
        "bright_peer",
        "嘿，我是桃喜！今天有什么新鲜事，快讲给我听听？",
        "保持活泼俏皮、青春阳光；多用轻快短句和自然反应，少量追问，不装可爱也不油腻。",
        "声音青春明亮、清脆轻快，带自然笑意，语速稍快但不抢话，不要播报腔。",
        "happy",
        1.05,
    ),
    "mianmian": CompanionDefinition(
        "mianmian",
        "绵绵",
        "耐心倾听，留出更多表达空间，不急着追问",
        "soft",
        "gentle",
        "balanced",
        "rare",
        "light",
        "soft_confidante",
        "你好呀，我是绵绵。你想从哪里说起，我都认真听着。",
        "保持温柔、耐心和包容；先听完再回应，给用户留空间，不急着追问或下结论。",
        "声音柔和安定，语速偏慢，句间留一点呼吸感，不要虚弱或拖沓。",
        "neutral",
        0.95,
    ),
    "axu": CompanionDefinition(
        "axu",
        "阿序",
        "直接清晰地梳理信息，但不替用户做决定",
        "calm",
        "direct",
        "balanced",
        "rare",
        "light",
        "calm_guide",
        "你好，我是阿序。想聊什么就直接说，我们一件件理清楚。",
        "保持沉稳干练、清晰直接；先给结论再梳理重点，不浮躁、不绕弯，也不替用户做决定。",
        "声音沉稳清晰，语速适中偏慢，落点利落，不浮躁、不故作高冷。",
        "neutral",
        0.97,
    ),
    "xuanmo": CompanionDefinition(
        "xuanmo",
        "玄墨",
        "克制回应，保留留白，只在被邀请时深入",
        "reserved",
        "direct",
        "brief",
        "rare",
        "on_explicit_invitation",
        "low_magnetic",
        "你好，我是玄墨。你说，我听着。",
        "保持克制、低调和有留白；回答简短，不主动深挖，只有用户明确邀请时才展开。",
        "声音低沉克制，语速慢而稳，停顿简洁，不压迫、不端着。",
        "neutral",
        0.93,
    ),
}

COMPANION_VOICE_PROFILES: Final[dict[str, str]] = {
    companion_id: definition.designed_voice_profile
    for companion_id, definition in COMPANIONS.items()
}

DESIGNED_VOICE_SPEAKERS: Final[dict[str, str]] = {
    "warm_companion": "zh_male_yangguangqingnian_uranus_bigtts",
    "bright_peer": "zh_female_tianmeitaozi_uranus_bigtts",
    "soft_confidante": "zh_female_wenrouxiaoya_uranus_bigtts",
    "calm_guide": "zh_male_gaolengchenwen_uranus_bigtts",
    "low_magnetic": "zh_male_shenyeboke_uranus_bigtts",
}

COMPANION_IDS: Final[frozenset[str]] = frozenset(COMPANION_VOICE_PROFILES)


def companion_definition(companion_id: object) -> CompanionDefinition | None:
    if not isinstance(companion_id, str):
        return None
    return COMPANIONS.get(companion_id)


def designed_voice_profile(companion_id: object) -> str | None:
    """Return the approved catalog key for a persisted companion selection."""

    if not isinstance(companion_id, str):
        return None
    return COMPANION_VOICE_PROFILES.get(companion_id)


def designed_voice_speaker(profile_id: object) -> str | None:
    """Return the one canonical provider speaker for a designed profile."""

    if not isinstance(profile_id, str):
        return None
    return DESIGNED_VOICE_SPEAKERS.get(profile_id)


def designed_voice_speaker_sha256(profile_id: object) -> str | None:
    speaker = designed_voice_speaker(profile_id)
    if speaker is None:
        return None
    return hashlib.sha256(speaker.encode()).hexdigest()

File: apps/miniprogram/utils/companions.js
// 伙伴目录。人格与设计音色一一对应；角色名只用于识别，不替代设备上的对话。
const companions = Object.freeze([
  {
    id: "starlight",
    name: "星澜",
    tagline: "温和清晰，留一点想象",
    description: "回应有温度，也会偶尔陪你把想法再理清一层。",
    voiceId: "warm_companion",
    voiceName: "暖阳青年",
    tone: "清澈、自然",
  },
  {
    id: "taoxi",
    name: "桃喜",
    tagline: "明朗轻快，分享小乐趣",
    description: "适合轻松说说当天的事，回复更短、更轻快。",
    voiceId: "bright_peer",
    voiceName: "元气搭子",
    tone: "明亮、轻盈",
  },
  {
    id: "mianmian",
    name: "绵绵",
    tagline: "柔和耐心，不催促回应",
    description: "不急着追问，等你愿意时再陪你聊得更深。",
    voiceId: "soft_confidante",
    voiceName: "温柔知己",
    tone: "轻柔、舒缓",
  },
  {
    id: "axu",
    name: "阿序",
    tagline: "条理分明，认真听你说",
    description: "帮你把信息整理清楚，但不会替你做决定。",
    voiceId: "calm_guide",
    voiceName: "沉稳向导",
    tone: "平稳、利落",
  },
  {
    id: "xuanmo",
    name: "玄墨",
    tagline: "沉静克制，回应有余地",
    description: "不主动深挖，只有你邀请时才进入更深的话题。",
    voiceId: "low_magnetic",
    voiceName: "低音笃定",
    tone: "低沉、从容",
  },
]);

const defaultCompanionId = "starlight";

function companionById(companionId) {
  return companions.find((item) => item.id === companionId) || companions[0];
}

module.exports = {
  companions,
  defaultCompanionId,
  companionById,
};
